### DISCUSSION TITLE
Error while training ner and re simultaneously: (ValueError: operands could not be broadcast together with shapes)

### DISCUSSION THREAD

**Discussion by TarkanKaracay** [explosion/spaCy#13142] [Help: Coding & Implementations]
Hello,
I'm trying to train named entity recognition and relation extractor same time with transformers.

This is my config file. Is there anything wrong or missing?

> [paths]
> train = "/content/drive/MyDrive/sil/trouCY_relations_training.spacy"
> dev = "/content/drive/MyDrive/sil/trouCY_relations_dev.spacy"
> vectors = null
> init_tok2vec = null
> 
> [system]
> gpu_allocator = "pytorch"
> seed = 0
> 
> [nlp]
> lang = "en"
> pipeline = ["transformer","ner", "relation_extractor"]
> tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}
> disabled = []
> before_creation = null
> after_creation = null
> after_pipeline_creation = null
> batch_size = 128
> 
> [components]
> 
> [components.ner]
> factory = "ner"
> incorrect_spans_key = null
> moves = null
> scorer = {"@scorers":"spacy.ner_scorer.v1"}
> update_with_oracle_cut_size = 100
> 
> [components.ner.model]
> @architectures = "spacy.TransitionBasedParser.v2"
> state_type = "ner"
> extra_state_tokens = false
> hidden_width = 64
> maxout_pieces = 2
> use_upper = false
> nO = null
> 
> [components.ner.model.tok2vec]
> @architectures = "spacy-transformers.Tok2VecTransformer.v3"
> name = "dmis-lab/biobert-base-cased-v1.2"
> mixed_precision = false
> pooling = {"@layers":"reduce_mean.v1"}
> grad_factor = 1.0
> 
> [components.ner.model.tok2vec.get_spans]
> @span_getters = "spacy-transformers.strided_spans.v1"
> window = 128
> stride = 96
> 
> [components.ner.model.tok2vec.grad_scaler_config]
> 
> [components.ner.model.tok2vec.tokenizer_config]
> use_fast = true
> 
> [components.ner.model.tok2vec.transformer_config]
> 
> [components.transformer]
> factory = "transformer"
> max_batch_items = 4096
> set_extra_annotations = {"@annotation_setters":"spacy-transformers.null_annotation_setter.v1"}
> 
> [components.transformer.model]
> @architectures = "spacy-transformers.TransformerModel.v3"
> name = "dmis-lab/biobert-base-cased-v1.2"
> mixed_precision = false
> 
> [components.transformer.model.get_spans]
> @span_getters = "spacy-transformers.strided_spans.v1"
> window = 128
> stride = 96
> 
> [components.transformer.model.grad_scaler_config]
> 
> [components.transformer.model.tokenizer_config]
> use_fast = true
> 
> [components.transformer.model.transformer_config]
> 
> 
> [components.relation_extractor]
> factory = "relation_extractor"
> threshold = 0.5
> 
> [components.relation_extractor.model]
> @architectures = "rel_model.v1"
> 
> [components.relation_extractor.model.create_instance_tensor]
> @architectures = "rel_instance_tensor.v1"
> 
> [components.relation_extractor.model.create_instance_tensor.tok2vec]
> @architectures = "spacy-transformers.TransformerListener.v1"
> grad_factor = 1.0
> 
> [components.relation_extractor.model.create_instance_tensor.tok2vec.pooling]
> @layers = "reduce_mean.v1"
> 
> [components.relation_extractor.model.create_instance_tensor.pooling]
> @layers = "reduce_mean.v1"
> 
> [components.relation_extractor.model.create_instance_tensor.get_instances]
> @misc = "rel_instance_generator.v1"
> max_length = 10
> 
> [components.relation_extractor.model.classification_layer]
> @architectures = "rel_classification_layer.v1"
> nI = null
> nO = null
> 
> 
> 
> [corpora]
> 
> [corpora.train]
> @readers = "spacy.Corpus.v1"
> path = ${paths.train}
> max_length = 0
> 
> [corpora.dev]
> @readers = "spacy.Corpus.v1"
> path = ${paths.dev}
> max_length = 0
> 
> [training]
> seed = ${system.seed}
> gpu_allocator = ${system.gpu_allocator}
> dropout = 0.1
> accumulate_gradient = 1
> patience = 1600000
> max_epochs = 0
> max_steps = 1000
> eval_frequency = 10
> frozen_components = []
> annotating_components = ["ner"]
> dev_corpus = "corpora.dev"
> train_corpus = "corpora.train"
> before_to_disk = null
> logger = {"@loggers":"spacy.ConsoleLogger.v1"}
> 
> [training.batcher]
> @batchers = "spacy.batch_by_padded.v1"
> discard_oversize = true
> size = 2000
> buffer = 256
> 
> [training.optimizer]
> @optimizers = "Adam.v1"
> beta1 = 0.9
> beta2 = 0.999
> L2_is_weight_decay = true
> L2 = 0.01
> grad_clip = 1.0
> use_averages = false
> eps = 0.00000001
> 
> [training.optimizer.learn_rate]
> @schedules = "warmup_linear.v1"
> warmup_steps = 250
> total_steps = 20000
> initial_rate = 5e-5
> 
> [training.score_weights]
> rel_micro_p = 0.0
> rel_micro_r = 0.0
> rel_micro_f = 1.0
> 
> 
> [pretraining]
> 
> [initialize]
> vectors = ${paths.vectors}
> init_tok2vec = ${paths.init_tok2vec}
> vocab_data = null
> lookups = null
> before_init = null
> after_init = null
> 
> [initialize.components]
> 
> [initialize.tokenizer]

My custom_functions.py file is simply a copy-paste of the contents of rel_pipe.py and rel_model.py.
When I use this code I get an error:

> !python -m spacy train -g 0 /content/drive/MyDrive/sil/config_spacy.cfg --output /content/drive/MyDrive/sil/ -c /content/drive/MyDrive/sil/custom_functions.py

ERROR
> 2023-11-21 11:53:00.118526: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
> 2023-11-21 11:53:00.118582: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
> 2023-11-21 11:53:00.118612: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
> 2023-11-21 11:53:01.244493: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
> ℹ Saving to output directory: /content/drive/MyDrive/sil
> ℹ Using GPU: 0
> 
> =========================== Initializing pipeline ===========================
> ✔ Initialized pipeline
> 
> ============================= Training pipeline =============================
> ℹ Pipeline: ['transformer', 'ner', 'relation_extractor']
> ℹ Set annotations on update for: ['ner']
> ℹ Initial learn rate: 0.0
> E    #       LOSS TRANS...  LOSS NER  LOSS RELAT...  ENTS_F  ENTS_P  ENTS_R  REL_MICRO_P  REL_MICRO_R  REL_MICRO_F  SCORE 
> ---  ------  -------------  --------  -------------  ------  ------  ------  -----------  -----------  -----------  ------
> ⚠ Aborting and saving the final best model. Encountered exception:
> ValueError('operands could not be broadcast together with shapes (25750, 5)
> (498, 5)')
> Traceback (most recent call last):
>   File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
>     exec(code, run_globals)
>   File "/usr/local/lib/python3.10/dist-packages/spacy/__main__.py", line 4, in <module>
>     setup_cli()
>   File "/usr/local/lib/python3.10/dist-packages/spacy/cli/_util.py", line 87, in setup_cli
>     command(prog_name=COMMAND)
>   File "/usr/local/lib/python3.10/dist-packages/click/core.py", line 1157, in __call__
>     return self.main(*args, **kwargs)
>   File "/usr/local/lib/python3.10/dist-packages/typer/core.py", line 778, in main
>     return _main(
>   File "/usr/local/lib/python3.10/dist-packages/typer/core.py", line 216, in _main
>     rv = self.invoke(ctx)
>   File "/usr/local/lib/python3.10/dist-packages/click/core.py", line 1688, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/usr/local/lib/python3.10/dist-packages/click/core.py", line 1434, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/usr/local/lib/python3.10/dist-packages/click/core.py", line 783, in invoke
>     return __callback(*args, **kwargs)
>   File "/usr/local/lib/python3.10/dist-packages/typer/main.py", line 683, in wrapper
>     return callback(**use_params)  # type: ignore
>   File "/usr/local/lib/python3.10/dist-packages/spacy/cli/train.py", line 54, in train_cli
>     train(config_path, output_path, use_gpu=use_gpu, overrides=overrides)
>   File "/usr/local/lib/python3.10/dist-packages/spacy/cli/train.py", line 84, in train
>     train_nlp(nlp, output_path, use_gpu=use_gpu, stdout=sys.stdout, stderr=sys.stderr)
>   File "/usr/local/lib/python3.10/dist-packages/spacy/training/loop.py", line 135, in train
>     raise e
>   File "/usr/local/lib/python3.10/dist-packages/spacy/training/loop.py", line 118, in train
>     for batch, info, is_best_checkpoint in training_step_iterator:
>   File "/usr/local/lib/python3.10/dist-packages/spacy/training/loop.py", line 220, in train_while_improving
>     nlp.update(
>   File "/usr/local/lib/python3.10/dist-packages/spacy/language.py", line 1193, in update
>     proc.update(examples, sgd=None, losses=losses, **component_cfg[name])  # type: ignore
>   File "/content/drive/MyDrive/sil/custom_functions.py", line 133, in update
>     loss, gradient = self.get_loss(examples, predictions)
>   File "/content/drive/MyDrive/sil/custom_functions.py", line 146, in get_loss
>     gradient = scores - truths
>   File "cupy/_core/core.pyx", line 1271, in cupy._core.core._ndarray_base.__sub__
>   File "cupy/_core/_kernel.pyx", line 1251, in cupy._core._kernel.ufunc.__call__
>   File "cupy/_core/internal.pyx", line 382, in cupy._core.internal._broadcast_core
> ValueError: operands could not be broadcast together with shapes (25750, 5) (498, 5)

Do you have any ideas to solve the problem?

**Comment by rmitsch** [comment]
Hi @TarkanKaracay, does training only NER without REL work?

**Reply by TarkanKaracay** [reply]
Yes. I trained NER without REL. It perfectly worked.
